src/snappy_device_agents/devices/maas2/maas2.py

```diff
@@ -72,7 +75,12 @@ def provision(self):
         distro = provision_data.get("distro", "xenial")
         kernel = provision_data.get("kernel")
         user_data = provision_data.get("user_data")
-        self.deploy_node(distro, kernel, user_data)
+        storage_data = provision_data.get("disks")
+        if storage_data:
+            maas_storage = MaasStorage(
```

Comment:
self.maas_storage?

Reply:
Addressed in: https://github.com/canonical/snappy-device-agents/pull/60#discussion_r1244557031

Reply:
> Is there a piece of this that also needs to always ensure we reset the disk back to flat? regardless of there being storage data or not

Should test this... my understanding was that this was done on a per deployment basis (e.g. "Hey maas, deploy X but make the partitions look like $LAYOUT this time") and then the next time MAAS would go back to whatever the default is in the machine config in MAAS... but we definitely want to confirm that is indeed the case before this goes live and starts accidentally modifying things :D

Reply:
OK, over our stand-up we talked about this, and this does, indeed, permanently alter the config in MAAS... so what needs to happen during any test run that calls this module is:

1. the module obtains the current storage state BEFORE the machine is acquired, and modified
2. the module does whatever its told to via the configuration presented (e.g. it reconfigures storage based on Alex's needs)
3. at the end, the module restores the initial state back.

This is crucial as we have no standard config, so some machines may have a single disk, in which case telling MAAS to just "put it back as a flat layout" is fine" but some machines have multiple disks that are all mounted to unique mount points for testing purposes, in which case "put it back as a flat layout" is not enough as that will only restore the single bood drive, it won't reconfigure the other drives...

The extreme example of this would be using hoggus, which has 14 drives, all mounted to unique mount points, so we would want to preserve those mount points.

Also, noticed from testing that drives have been arbitrarily renamed. Unless there's a need to rename drives for testing purposes, that shouldn't happen at all (just for sanity sake), OR, part of that statefulness would need to be restoring the original drive names too if there's a use case for renaming the drives during a test run.  I think Adrian said it wasn't necessary really to change though, so that could be removed.

Reply:
So now we will:
- Parse storage configuration via node block-devices read (done already)
- Capture in a dict
- Clear the storage configuration
- Write the Testflinger disk layout
- Clear the storage configuration
- Write the default storage configuration captured at the start.

Reply:
the only concern is that say something happens and the last task doesn't get run; ie something failed and "write the default storage configuration captured at start" doesn't get run. Now you are in a state where the previous config is the default config, because the reset failed.

I would expect we can make some assumption like
* the default disk config is always flat (if it is different it can be specified in the test)
* always restore to flat at the start of a test, incase something went wrong during a previous test

there could be an argument made about changing the disk configuration back after the test too, since you have to release the node. maybe someone reserved the node, or wants to inspect the node after the test.

Reply:
My concern isn't just that the machine's disk config is changed permanently... it's the unused disks that worry me. We do not have uniform config. As you know, we have servers that run the range from 1 disk to at least 14 individual disks.  For SRU all the disks in a machine have to be mounted in some way.  Unfortunately, just saying "Return the machine to a flat layout" doesn't restore those extra disks to their previous state.

For example, this is wildorage with 4 disks, 1 is a boot disk, the other three are partitioned and mounted as a data drives for test purposes:
![Screenshot from 2023-06-28 17-24-08](https://github.com/canonical/snappy-device-agents/assets/4295496/3d925d89-127f-413a-87b7-08321d4184d9)

Then lets change that to a simple bcache layout, simulating a test run someone launches that needs bcache:
![Screenshot from 2023-06-28 17-25-21](https://github.com/canonical/snappy-device-agents/assets/4295496/322312c8-9646-4b00-8309-d3feeb873c3a)

Now at the end of the run, if we simply request "reset to a flat layout" we do get a flat layout, but only on the first disk, the remaining ones are not restored.
![Screenshot from 2023-06-28 17-25-21](https://github.com/canonical/snappy-device-agents/assets/4295496/ba20e1a2-a9ef-4988-ab6d-c94033ab0652)

I don't know what needs to happen to preserve that state, but I don't want the default configs (which are unique to each machine, unfortunately) to be lost with each test run.

Reply:
IN a perfect world it would be as I had imagined it to be where you could tell maas to TEMPORARILY alter the disk layout for THIS deployment, while preserving the existing config for subsequent deployments... but it appears that is not something MAAS is capable of, as far as I can tell.  I'll ask them though, just to be sure.

Reply:
Reseting the disk layout will take place in the Testflinger cleanup phase, which is decoupled from the job itself.

Reply:
what if the agent dies or is updated mid test, or network goes down?

would it be possible to state the default disk configuration in the agent config, so if no one specifies one, it uses the state posted there?
then you dont have any possible race conditions

Reply:
It sounds like that would be difficult to achieve after speaking with Jeff about this earlier.

What we could do is write the layout to a file on the agent and then that could be used to reset the layout outside of this context if necessary. I could utilize the agent state monitoring I'm doing with Docker (for healthchecks, for example) - to check and reset this layout (if required) once the agent enters a waiting state.

I'll integrate writing the layout to a file on the agent into this round of changes... Thoughts?

Reply:
This would also keep a disk layout log of sorts to keep the state over time

Reply:
This of course also has some issues if for some reason the file does not exist, and mapping this to non-ephemeral storage on the Docker host would be key. In this case manual intervention would be necessary, but it sounds like that is acceptable if absolutely required and the goal would be to minimize the instances in which manual reconfiguration would be necessary.